internal/receiver/smartagentreceiver/convert.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2021, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package smartagentreceiver
+
+import (
+	"fmt"
+
+	sfx "github.com/signalfx/golib/v3/datapoint"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+)
+
+var (
+	errUnsupportedMetricTypeTimestamp = fmt.Errorf("unsupported metric type timestamp")
+	errNoIntValue                     = fmt.Errorf("no valid value for expected IntValue")
+	errNoFloatValue                   = fmt.Errorf("no valid value for expected FloatValue")
+)
+
+type Converter struct {
+	logger *zap.Logger
+}
+
+// Based on https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.15.0/receiver/signalfxreceiver/signalfxv2_to_metricdata.go
+func (c *Converter) toMetrics(datapoints []*sfx.Datapoint) (pdata.Metrics, int) {
+	numDropped := 0
+	md := pdata.NewMetrics()
+	md.ResourceMetrics().Resize(1)
+	rm := md.ResourceMetrics().At(0)
+
+	rm.InstrumentationLibraryMetrics().Resize(1)
+	ilm := rm.InstrumentationLibraryMetrics().At(0)
+
+	metrics := ilm.Metrics()
+	metrics.Resize(len(datapoints))
+
+	i := 0
+	for _, datapoint := range datapoints {
+		if datapoint == nil {
+			continue
+		}
+
+		m := metrics.At(i)
+		err := setDataType(datapoint, m)
+		if err != nil {
+			numDropped++
+			c.logger.Debug("SignalFx datapoint type conversion error",
+				zap.Error(err),
+				zap.String("metric", datapoint.String()))
+			continue
+		}
+
+		m.SetName(datapoint.Metric)
+
+		switch m.DataType() {
+		case pdata.MetricDataTypeIntGauge:
+			err = fillIntDatapoint(datapoint, m.IntGauge().DataPoints())
+		case pdata.MetricDataTypeIntSum:
+			err = fillIntDatapoint(datapoint, m.IntSum().DataPoints())
+		case pdata.MetricDataTypeDoubleGauge:
+			err = fillDoubleDatapoint(datapoint, m.DoubleGauge().DataPoints())
+		case pdata.MetricDataTypeDoubleSum:
+			err = fillDoubleDatapoint(datapoint, m.DoubleSum().DataPoints())
+		}
+
+		if err != nil {
+			numDropped++
+			c.logger.Debug("SignalFx datapoint datum conversion error",
+				zap.Error(err),
+				zap.String("metric", datapoint.Metric))
+			continue
+		}
+
+		i++
+	}
+
+	metrics.Resize(i)
+
+	return md, numDropped
+
+}
+func setDataType(datapoint *sfx.Datapoint, m pdata.Metric) error {
+	sfxMetricType := datapoint.MetricType
+	if sfxMetricType == sfx.Timestamp {
+		return errUnsupportedMetricTypeTimestamp
+	}
+
+	var isFloat bool
+	switch datapoint.Value.(type) {
+	case sfx.IntValue:
+		isFloat = false
```

Comment:
nit: this could be omitted since booleans are by default set to `false`.

Reply:
Still not used to go's lack of falling through in switches.  Removed.